File: udf-examples/src/main/java/com/nvidia/spark/rapids/udf/hive/URLDecode.java
/*
 * Copyright (c) 2020-2021, NVIDIA CORPORATION.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.nvidia.spark.rapids.udf.hive;

import ai.rapids.cudf.ColumnVector;
import ai.rapids.cudf.DType;
import ai.rapids.cudf.Scalar;
import com.nvidia.spark.RapidsUDF;
import org.apache.hadoop.hive.ql.exec.UDF;

import java.io.UnsupportedEncodingException;
import java.net.URLDecoder;

/**
 * A Hive user-defined function (UDF) that decodes URL-encoded strings.
 * This class demonstrates how to implement a simple Hive UDF that also
 * provides a RAPIDS implementation that can run on the GPU when the query
 * is executed with the RAPIDS Accelerator for Apache Spark.
 */
public class URLDecode extends UDF implements RapidsUDF {

  /** Row-by-row implementation that executes on the CPU */
  public String evaluate(String s) {
    String result = null;
    if (s != null) {
      try {
        result = URLDecoder.decode(s, "utf-8");
      } catch (IllegalArgumentException ignored) {
        result = s;
      } catch (UnsupportedEncodingException e) {
        // utf-8 is a builtin, standard encoding, so this should never happen
        throw new RuntimeException(e);
      }
    }
    return result;
  }

  /** Columnar implementation that runs on the GPU */
  @Override
  public ColumnVector evaluateColumnar(ColumnVector... args) {
    // The CPU implementation takes a single string argument, so similarly
    // there should only be one column argument of type STRING.
    if (args.length != 1) {
      throw new IllegalArgumentException("Unexpected argument count: " + args.length);
    }
    ColumnVector input = args[0];
    if (!input.getType().equals(DType.STRING)) {
      throw new IllegalArgumentException("Argument type is not a string column: " +
          input.getType());
    }

    // The cudf urlDecode does not convert '+' to a space, so do that as a pre-pass first.
    // All intermediate results are closed to avoid leaking GPU resources.
    try (Scalar plusScalar = Scalar.fromString("+");
         Scalar spaceScalar = Scalar.fromString(" ");
         ColumnVector replaced = input.stringReplace(plusScalar, spaceScalar)) {
      return replaced.urlDecode();
    }
  }
}

File: tests/src/test/scala/com/nvidia/spark/rapids/tests/udf/scala/URLDecode.scala
/*
 * Copyright (c) 2021-2022, NVIDIA CORPORATION.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.nvidia.spark.rapids.tests.udf.scala

import java.net.URLDecoder

import ai.rapids.cudf.{ColumnVector, DType, Scalar}
import com.nvidia.spark.RapidsUDF

/**
 * A Scala user-defined function (UDF) that decodes URL-encoded strings.
 * This class demonstrates how to implement a Scala UDF that also
 * provides a RAPIDS implementation that can run on the GPU when the query
 * is executed with the RAPIDS Accelerator for Apache Spark.
 */
class URLDecode extends Function[String, String] with RapidsUDF with Serializable {
  /** Row-by-row implementation that executes on the CPU */
  override def apply(s: String): String = {
    Option(s).map { s =>
      try {
        URLDecoder.decode(s, "utf-8")
      } catch {
        case _: IllegalArgumentException => s
      }
    }.orNull
  }

  /** Columnar implementation that runs on the GPU */
  override def evaluateColumnar(numRows: Int, args: ColumnVector*): ColumnVector = {
    // The CPU implementation takes a single string argument, so similarly
    // there should only be one column argument of type STRING.
    require(args.length == 1, s"Unexpected argument count: ${args.length}")
    val input = args.head
    require(numRows == input.getRowCount, s"Expected $numRows rows, received ${input.getRowCount}")
    require(input.getType == DType.STRING, s"Argument type is not a string: ${input.getType}")

    // The cudf urlDecode does not convert '+' to a space, so do that as a pre-pass first.
    // All intermediate results are closed to avoid leaking GPU resources.
    val plusScalar = Scalar.fromString("+")
    try {
      val spaceScalar = Scalar.fromString(" ")
      try {
        val replaced = input.stringReplace(plusScalar, spaceScalar)
        try {
          replaced.urlDecode()
        } finally {
          replaced.close()
        }
      } finally {
        spaceScalar.close()
      }
    } finally {
      plusScalar.close()
    }
  }
}
